Initial implemetation

smarnach opened the pull request:
This is an initial implementation for the Ansible role to deploy a load balancing server for the OpenCraft Instance Manager.  The README.md file contains some basic information.
#### Testing instructions
1. Deploy an OpenStack instance in the region GRA1 using the account "OVH - OpenStack - Horizon - OpenCraft IM - Dev".  Use the image "xenial-server-cloudimg-unmodified-20160517".  You probably need to upload an SSH key before deploying the instance.
2. Check out the test playbook in [this branch](https://github.com/open-craft/deployment-deploy-services/tree/smarnach/lb-test).  (First check out the master version of [deployment-secrets-services](https://github.com/open-craft/deployment-secrets-services) if you haven't already; then go the `deploy` subdirectory and check out the `smarnach/lb-test` branch.)
3. In the top directory of the deployment-secrets-services repository, edit the Ansible inventory `hosts and add
   
   ```
   [load-balancer]
   haproxy-test.opencraft.hosting ansible_host=1.2.3.4
   ```
   
   Use the IP address of your newly deployed VM from step 1.
4. From the same directory, run the playbook using
   
   ```
   ansible-playbook -vvv deploy/deploy-lb.yml -u ubuntu --private-key <your-ssh-id> \
       -i hosts -l haproxy-test.opencraft.hosting -e LETSENCRYPT_USE_STAGING=true
   ```

The playbook should run successfully.  Once it's finished, add a backend as desribed in `README.md`.  A simple haproxy backend definition is to forward to some webserver running on localhost, e.g.

```
backend be_pydoc
    server srv_pydoc localhost:8000
```

Then run `pydoc -p 8000` in another terminal.

smarnach committed:
Initial implemetation

smarnach committed:
Make some paths configurable via Ansible variables.

smarnach committed:
Add locks to protect against half-written config files.

smarnach committed:
Restart configuration file watcher when the playbook is run.

smarnach committed:
Block on the locks for a second when reloading haproxy.

smarnach committed:
Add certificate management script.

smarnach committed:
Update README to reflect automatic certificate management.

smarnach committed:
Add Let's Encrypt fake certificate when testing.

smarnach committed:
Add cron job to run certificate management script every 5 minutes.

smarnach committed:
Use correct Let's Encrypt staging root certificate.

bdero commented on tasks/main.yml:
```diff
@@ -0,0 +1,104 @@
+  command: systemctl {{ item }} haproxy-config-watcher
+  with_items:
+    - enable
+    - restart
```
Nit: Which Ansible version range is this playbook looking to support? 2.2+ has a systemd module: https://docs.ansible.com/ansible/systemd_module.html

smarnach replied:
We are [currently using Ansible 2.0.1](https://github.com/open-craft/deployment-deploy-services/blob/master/requirements.txt).  While it's probably possible to upgrade to 2.2 without any issues, I don't want to do this as part of a story that already got too big.

bdero commented:
Unfortunately, I am unable to import my public key into that OVH account. Key import worked on a different account. :/

![](http://i.imgur.com/RIniaGS.png)

CC @antoviaque: Is there something I'm missing here?

antoviaque commented:
@bdero I don't see anything wrong on the screenshot - it's too bag the error message from horizon isn't more explicit. Maybe try again from the CLI, you might get more? Maybe the format or length of your key isn't supported?

antoviaque commented:
Although that wouldn't have worked on another account of it was the key format. Maybe a quota? I would still try the CLI, and regenerating a new key just in case.

bdero commented on templates/haproxy-reload:
```diff
@@ -0,0 +1,27 @@
+HAPROXY_RELOAD_FILE="{{ LOAD_BALANCER_RELOAD_FILE }}"
+
+if [ -f "$HAPROXY_RELOAD_FILE" -o "$1" = "-f" ]; then
+    rm "$HAPROXY_RELOAD_FILE"
```
This fails when the `$HAPROXY_RELOAD_FILE` doesn't exist but the `-f` flag has been passed (happens on the last step of the role).

To get around this temporarily I just surrounded the line with it's own file check:

``` sh
if [ -f "$HAPROXY_RELOAD_FILE" ]; then
    rm "$HAPROXY_RELOAD_FILE"
fi
```

smarnach replied:
Thanks, you are right.  I didn't run into this since usually the file will exist in the last step of the role.  Only if the cron job happens to have run earlier it will be already deleted.  The easiest fix is to use

```
rm -f "$HAPROXY_RELAOD_FILE"
```

which won't complain if the file does not exist.

(In theory, the last step of the playbook is redeundant, since the cron job will take care of restarting the service.  However, this seems to implicit to me, and might break silently if the tasks in the role are reordered or refactored.)

bdero reviewed (changes requested):
I left a couple of comments.

I tested this out and the haproxy balancing and related crons work. I think I need more details on the letsencrypt setup. Out of the box, haproxy redirected to https, but did not supply my browsers an SSL cert. No evidence of issues via the logs.

Let me know if there's something more I should be expecting out of `letsencrypt` here.

smarnach commented:
@bdero The server should take care of retrieving and installing a certificate all by itself if you follow the steps in the README.md file (in particular, you need to point a DNS record to the server and wait for DNS changes to propagate).  If you configure a valid backend, you should be able to access the backend via your browser.

When testing, you are probably using `LETSENCRYPT_USE_STAGING=true`, so the certificate won't be accepted by default by the browser, and you will have to add an exception.

I suspect the problem is with the heuristic to determine the server's own external IP address.  There simply is no reliable way to determine the external IP address, so I should investigate ways of passing it in via Ansible.  This might require me to write a custom Ansible module, or alternatively to require the IP address to be passed in in Ansible variable.

smarnach committed:
Configure certificate management via a configuration file and minor fixes.

Fixes include:

* use "rm -f" to delete file indicating that haproxy should be restarted.  This
  fixes an error that might occur when calling haproxy-reload with "-f".

* Resolve the IP of the server on the host running Ansible.  This is a much more
  robust way of obtaining the server IP.

smarnach committed:
Add logging to certificate management script.

smarnach committed:
Treat names resolving to s local IP address as having s valid DNS record.

smarnach commented:
@bdero I've addressed your comments, introduced a saner way of configuring the manage_certs.py script and added logging.  Please take another look!

I supsect the problem you ran into was caused by the broken way I used to detect whether the DNS record of a domain points to the haproxy server.  I'm now using a much saner approach to determine the public IP address of the server.  However, this approach requires that `dnspython` be installed in the virtualenv used to run Ansible, so you need to do `pip install dnspython`.

I also suggest you add `-e LOAD_BALANCER_MANAGE_CERTS_LOG_LEVEL=debug` to the `ansible-playbook` command line to get verbose logging of the certifcate management script.  The log messages are send to syslog for simplicity, and can be inspected using `journalctl -u cron -e` (or by `less /var/log/syslog`).

bdero commented on defaults/main.yml:
```diff
@@ -0,0 +1,26 @@
+LOAD_BALANCER_SERVER_IP: "{{ lookup('dig', ansible_host) }}"
```
We'll need to add `dnspython` to `requirements.txt` in deployment-secrets-services. (Maybe this has already been addressed)

smarnach replied:
@bdero Yes, I've already done this locally, but thanks for reminding me.  It will be included in the PR with the actual playbook.

bdero commented:
Thanks @smarnach - I went ahead and tested this again and found one quirk with the lookup, but everything else looks good!

My hosts file looked like this:

```
[load-balancer]
haproxy-test.opencraft.hosting ansible_host=149.202.168.38
```

I ran this:

```
ansible-playbook -vvv deploy/deploy-lb.yml -u ubuntu --private-key ~/.ssh/id_rsa -i hosts -l haproxy-test.opencraft.hosting -e LETSENCRYPT_USE_STAGING=true -e LOAD_BALANCER_MANAGE_CERTS_LOG_LEVEL=debug
```

Here's how the `manage_certs.conf` template rendered:

```
--server-ip NXDOMAIN
--haproxy-backend-map /etc/haproxy/backend.map
--haproxy-certs-dir /etc/haproxy/certs
--contact-email ops@example.com
--log-level debug
--letsencrypt-use-staging
--letsencrypt-fake-cert /etc/ssl/certs/letsencrypt-fake-cert.crt
```

And the logs pointed out the mismatch:

```
Oct 08 02:30:02 haproxy-test.opencraft.hosting manage_certs.py[21757]: The DNS record for the domain haproxy-test.bdero.me does not point to this server.
Oct 08 02:30:02 haproxy-test.opencraft.hosting manage_certs.py[21757]: This server does not have a valid cert for the domain haproxy-test.bdero.me.
```

So it looks like a quirk of this is that, while it otherwise works, `dns.resolver` in `dnspython` will still try to do an actual DNS query even if you pass it an ipv4 address.

Swapping out the address for the domain I made an A record for resolved the issue:

```
[load-balancer]
haproxy-test.opencraft.hosting ansible_host=haproxy-test.bdero.me
```

```
--server-ip 149.202.168.38
...
```

If we need to support IPs in host files, we should be able to resolve the address with something like this:

```
LOAD_BALANCER_SERVER_IP: "{{ ansible_host | ipaddr or lookup('dig', ansible_host) }}"
```

Note: My jinja2 operator precedence might be way off here. Also not sure if filters are allowed to be used this way.

In the meantime, I think this looks good! :)

:+1:
- [x] I tested this: I deployed the load balancer to a test server and was able to route a domain using a backend with automatic certificate management. I was also able to test logging in both failure and success cases.
- [x] I read through the code
- [x] Includes documentation (the readme)

smarnach commented:
@bdero Thanks for the feedback!  I'm surprised about your problem with setting `ansible_host` to an IP address, since I've tested this as well and it works fine for me.  What version of dnspython did you use?  I'm using 1.15.0.

Could you also try the following playbook in your environment?

``` yaml
---
- hosts: localhost
  become: false
  gather_facts: false
  tasks:
  - debug:
      msg: "{{ lookup('dig', '131.252.210.176') }}"
```

You can run this in the same virtualenv using the command line

```
ansible-playbook -i localhost, test.yml
```

For me, this simply prints the IP address that's passed in.

antoviaque commented on templates/manage_certs.py:
```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""Certificate management for haproxy load balancing server.
```
Do you have in mind to create tests for this?

Specifically, I think integration tests will be important, both here within the context of the load balancer, and later on the other side in OpenCraft IM.

smarnach replied:
@antoviaque Yes, I would like to write integration tests for this script, as I mentioned on the ticket.  I would be grateful if we could exclude this from the current task, though.

bdero commented:
@smarnach Ah, I'm unable to reproduce the issue anymore. It may have just been left over config from the previous iteration that didn't end up getting overwritten. All seems well with the dnspython lookup method as it currently stands.

smarnach commented:
@bdero Thanks for checking!  I will merge this PR now.  It's the first version, so merging won't overwrite any previous version, and we can do any required further changes in additional, smaller, more manageable PRs.  I also need a commit hash to put in the requirements of the playbook.

bdero commented:
Thanks @smarnach. I agree, this PR was getting to be quite large. We should continue with this in smaller PRs from here.
